### PR TITLE
Feature: Runtime Content Filter

### PR DESCRIPTION
**Pull request by fdoxyz**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Implements the Runtime Content Filter feature described in [RFC 27](https://github.com/forem/rfcs/pull/27)

## Related Tickets & Documents

https://github.com/forem/rfcs/pull/27

## QA Instructions, Screenshots, Recordings

1. Boot app locally
2. Enable the `runtime_banner` Feature Flag
   - I like to use the console for this so `Flipper.enable(:runtime_banner)` would do the trick
4. Open any article (any page really) on your desktop browser 
   - You **shouldn't be able to see** a floating dark background banner with the text "Open with Forem app"
   - Check the `body` element and you should see `data-runtime` attribute that matches your environment. In macOS I get `Browser-macOS`
6. Open any article (any page really) using your local mobile device (DevTools mobile view won't work)
   - You **should be able to see** a floating dark background banner with the text "Open with Forem app"

### UI accessibility concerns?

No accessibility concerns for this PR. 

A follow up PR that will include the Runtime Banner (and also removing the Feature Flag) will require to be accessible though.

## Added tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)

The above ^ is a WIP

## [optional] What gif best describes this PR or how it makes you feel?

what do you see?

### CODE REVIEW

File: app/assets/javascripts/initializers/initializeRuntime.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+/* global Runtime */
+
+function initializeRuntime() {
+  document.body.dataset.runtime = Runtime.currentContext();
+}
```
Comment (fdoxyz) [inline app/assets/javascripts/initializers/initializeRuntime.js:5]: This is a tiny, one line initializer. I added it in a separate initializer file because it felt valuable to be separate concern, but looking back to it I'm not sure if that's true.

The alternative would be to simply add this one line to `app/assets/javascripts/initializePage.js`. I'm open to moving it there if others feel like that would be better.

File: app/assets/stylesheets/runtime-banner.scss [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  z-index: var(--z-sticky);
+
+  &__inner {
+    background-color: #0b0c0c;
+    color: #fff;
+    display: flex;
+    align-items: center;
+    justify-content: space-around;
+  }
+}
```
Comment (fdoxyz) [inline app/assets/stylesheets/runtime-banner.scss:15]: These are all placeholder values that will be fleshed out in a separate PR when implementing [RFC 72](https://github.com/forem/rfcs/pull/72). None of this is visible because of the `runtime_banner` Feature Flag

File: cypress/integration/loggedOutFlows/runtimeContext.spec.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  beforeEach(() => {
+    cy.testSetup();
+  });
+
+  it('should detect the current context and set data-runtime attr', () => {
+    // Visit the site
+    cy.visit('/');
+
+    // Check the body's data-runtime attribute
+    cy.get('body').should('have.attr', 'data-runtime');
```
Comment (fdoxyz) [inline cypress/integration/loggedOutFlows/runtimeContext.spec.js:11]: This test is very barebones. I originally wanted to make sure the attribute was set accordingly to the `navigator.platform` and `navigator.userAgent` values but I wasn't able to stub them. If this is okay we can carry on with this, otherwise it would be awesome if I get some help to make this work in Cypress 🙂

File: app/javascript/admin/controllers/sidebar_controller.js [hunk 4]
```diff
@@ -2,14 +2,14 @@ import { Controller } from 'stimulus';
 
 // eslint-disable-next-line no-restricted-syntax
 export default class SidebarController extends Controller {
-  static targets = [
-    'submenu'
-  ];
+  static targets = ['submenu'];
```
Comment (fdoxyz) [inline app/javascript/admin/controllers/sidebar_controller.js:5]: There's a few changes like this one added by the pre-commit hooks. I can get rid of them if someone feels like they shouldn't be included

File: app/assets/stylesheets/runtime.scss [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+.Browser-Windows,
+.Browser-Linux {
+  display: none;
+}
+
+// Elements with class 'ForemWebView' should be visible on any attribute that
+// matches that Medium regardless of OS
+body[data-runtime='ForemWebView-iOS'] .ForemWebView,
+body[data-runtime='ForemWebView-Android'] .ForemWebView {
+  display: block;
```
Comment (ludwiczakpawel) [inline app/assets/stylesheets/runtime.scss:33]: What happens if one of those elements should be `display: inline` or `display: flex` or anything else? In other words: to hide an element it's obvious to use `display: none` but to show the element we may need to use different values. 

I'm also not sure if that will ever be a problem, but I'm just thinking out loud.. 

Maybe we could use a variable like

```
body[data-runtime='ForemWebView-iOS'] .ForemWebView,
... {
  display: var(--runtime-display, block);
}
```

^ this will apply `block` value unless you specify otherwise:

```
.element { // <div class="element ForemWebView" />
  --runtime-display: flex;
}
```

File: app/assets/javascripts/initializers/runtime.js [hunk 6]
```diff
@@ -7,6 +7,62 @@
...
+      os = 'iOS';
+    } else if (windowsPlatforms.includes(window.navigator.platform)) {
+      os = 'Windows';
+    } else if (/Android/i.test(window.navigator.userAgent)) {
+      os = 'Android';
+    } else if (/Linux/i.test(window.navigator.platform)) {
+      os = 'Linux';
+    }
+
+    return os;
```
Comment (aitchiss) [inline app/assets/javascripts/initializers/runtime.js:63]: suggestion (not-blocking) - instead of reassigning the `os` variable we could just return inside the if statement -
```suggestion

    if (macosPlatforms.includes(window.navigator.platform)) {
      return 'macOS';
    } else if (iosPlatforms.includes(window.navigator.platform)) {
      return 'iOS';
    } else if (windowsPlatforms.includes(window.navigator.platform)) {
      return 'Windows';
    } else if (/Android/i.test(window.navigator.userAgent)) {
      return 'Android';
    } else if (/Linux/i.test(window.navigator.platform)) {
     return 'Linux';
    }

    return 'Unsupported';
```

File: app/javascript/admin/controllers/sidebar_controller.js (hunk 4 above)
Comment (aitchiss) [reply app/javascript/admin/controllers/sidebar_controller.js:5]: I don't feel strongly about it, but if it's simple enough to revert then I think it would help keep the scope of this PR clearer.

File: cypress/integration/loggedOutFlows/runtimeContext.spec.js (hunk 3 above)
Comment (aitchiss) [reply cypress/integration/loggedOutFlows/runtimeContext.spec.js:11]: I had a little look and it seems like we can stub them like this:

```js
 cy.visit('/', {
      onBeforeLoad: (win) => {
        Object.defineProperty(win.navigator, 'userAgent', {
          value: 'ForemWebView',
        });
        Object.defineProperty(win.navigator, 'platform', { value: 'Windows' });
      },
    });

    // Check the body's data-runtime attribute
    cy.get('body')
      .should('have.attr', 'data-runtime')
      .and('equal', 'ForemWebView-Windows');
```

It worked for me locally 🙂

**Review by aitchiss — COMMENTED** [review]
I tested this locally and it worked exactly as described! Left a suggestion, and what I think will unblock the Cypress test to verify the attribute's added properly. As we actually hide/show features with this filter it will be awesome to be able to add Cypress tests for that 🙂

File: app/assets/stylesheets/runtime.scss (hunk 5 above)
Comment (aitchiss) [reply app/assets/stylesheets/runtime.scss:33]: +1 to this suggestion - sounds good!

File: app/assets/javascripts/initializers/runtime.js [hunk 7]
```diff
@@ -7,6 +7,62 @@
...
+
+  /**
+   * This function returns a string that represents the current OS where the app
+   * is currently running. The currently supported Operating Systems are
+   * Windows, Linux, macOS, Android and iOS.
+   *
+   * @returns {String} One of the supported Operating Systems or 'Unsupported'
+   */
+  static currentOS() {
+    const macosPlatforms = ['Macintosh', 'MacIntel', 'MacPPC', 'Mac68K'];
```
Comment (rhymes) [inline app/assets/javascripts/initializers/runtime.js:46]: Does this include Apple M1?

File: app/javascript/crayons/MentionAutocompleteTextArea/__stories__/mention-autocomplete.md [hunk 8]
```diff
@@ -1,8 +1,12 @@
 ## Mention autocomplete
```
Comment (rhymes) [inline app/javascript/crayons/MentionAutocompleteTextArea/__stories__/mention-autocomplete.md:1]: did this slip in?

File: app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -106,6 +106,15 @@
     <%= yield %>
   </div>
 </div>
+<% if Flipper.enabled?(:runtime_banner) %>
```
Comment (rhymes) [inline app/views/layouts/application.html.erb:109]: non blocking: you can use the `FeatureFlag` object, which is a tiny wrapper over Flipper

**Review by rhymes — COMMENTED** [review]
I tested from my iPhone 8 with ngrok but I see the "Open with Forem app" notice on the homepage as well, not just on the article's page.

Works well in desktop.

See the screenshot:

qaMyH57w

File: app/assets/stylesheets/runtime.scss (hunk 5 above)
Comment (fdoxyz) [reply app/assets/stylesheets/runtime.scss]: This is pretty cool! I changed it to use this, thanks @ludwiczakpawel

File: cypress/integration/loggedOutFlows/runtimeContext.spec.js (hunk 3 above)
Comment (fdoxyz) [reply cypress/integration/loggedOutFlows/runtimeContext.spec.js]: I tried a few different ways using some parts of this suggestion but never really made it work. This works as a charm though, thank you @aitchiss! 

I just added as many as I could (i.e. we don't support `ForemWebView` in Android yet). Let me know if you think we should approach it differently 🙂

File: app/assets/javascripts/initializers/runtime.js (hunk 7 above)
Comment (fdoxyz) [reply app/assets/javascripts/initializers/runtime.js:46]: Funny enough [it does](https://forem-team.slack.com/archives/C7GE84DEJ/p1616596843134900?thread_ts=1616596026.134100&cid=C7GE84DEJ) 😄 

I added a test case for this, so if it ever changes we can update accordingly also in the test suite.

File: app/javascript/crayons/MentionAutocompleteTextArea/__stories__/mention-autocomplete.md (hunk 8 above)
Comment (fdoxyz) [reply app/javascript/crayons/MentionAutocompleteTextArea/__stories__/mention-autocomplete.md]: It did, from a pre-commit hook fixed some linter issues. Just reverted the changes 🙂

**Comment by fdoxyz** [comment]
> I tested from my iPhone 8 with ngrok but I see the "Open with Forem app" notice on the homepage as well, not just on the article's page.

Thanks for testing this out!

Yes, right now it should appear on every URL including the homepage. Ideally if someone shares any URL (including the root/homepage path) on a 3rd party app like Twitter (or if they open it using a mobile browser) then we should present the banner so they are able to open that URL on the mobile apps (or install them if they don't have them yet). The important part here is that the banner is only visible in certain Runtime contexts and hidden behind the Feature Flag.

The PR that will actually add the banner itself will come after we have this Runtime filtering available, so if it sounds good to you @rhymes I think we can ideally discuss if there are any exceptions to that rule in that follow up PR.

**Review by rhymes — APPROVED** [review]
Works well on my side! LGTM

**Review by aitchiss — APPROVED** [review]
Nice! It's cool to see as we implement features using this content filter, we'll be able to test in cypress that they show/hide on the right platforms 🙂
